### PR TITLE
fix(cross-model): read grok's structuredOutput and unwrap .text reviews

### PR DESCRIPTION
**Pull request by taihartman**
## Summary

A cross-model adversarial peer ran to completion on the `grok-cli` route, produced a complete
3-finding review, and the worker discarded all of it — logging `provider grok produced no usable
schema-shaped output; skipping fold-in`.

The cause is a naming mismatch. The grok-cli envelope names its parsed schema output
**`structuredOutput`**; `parse_structured` only probes `.structured_output`. Observed top-level keys
from a live grok-4.5 run:

```
text, stopReason, sessionId, requestId, thought, usage, num_turns,
total_cost_usd, total_cost_usd_ticks, modelUsage, structuredOutput
```

`jq 'has("structured_output")'` is `false` on that payload, and `.structuredOutput` held the complete
review.

This matters more than a normal parse miss because of what happens next. Stage 3d removes the
in-process `adversarial-reviewer` from the local roster the moment a peer job id comes back, and
Stage 5 forbids a late in-process retry. So the lens the skill routes to a different serving family
specifically to get independence **silently does not run**, and the outcome is classified
non-blocking and terminal. On `grok-cli` this is all-or-nothing by construction: `--json-schema`
forces buffered output, so there is no partial stream to salvage. Full cost, zero coverage, and the
Coverage section reports `peer ran, no usable output` — which understates it, because the content was
right there.

## Three failures stacked on one run

1. **`.structuredOutput` is never probed** (camelCase vs snake_case). This alone loses the review.
2. **`.text` carries the same review verbatim as a JSON string**, and grok emits an empty stub
   alongside the real object. jq accepts the pair as a *stream*, so an unslurped
   `jq -r '.text' | jq -e 'select(...)'` writes **both** objects into the output file and exits 0 —
   producing `JSONDecodeError: Extra data` downstream. Adding a naive `.text` branch is worse than
   adding none.
3. **`recover_findings_json` exits before it starts.** Its `'"findings"' not in txt` guard cannot
   match a key nested inside a JSON string, where it arrives escaped as `\"findings\"` — the
   character after `findings` is a backslash, not a quote. The safety net reports "nothing to
   recover" for a payload that fully contains a valid review.

## What changed

- Probe `.structuredOutput` alongside `.structured_output` in both workers.
- Slurp and shape-filter the `.text` stream, and prefer a populated review over an empty stub **in
  either order** — last-shaped-wins alone publishes the empty stub whenever it trails the real
  object, which reads downstream as "the peer found nothing."
- Widen the recovery guard to the bare word `findings`, and re-scan string values that could hold a
  wrapped review. The scan now collects candidates rather than tracking a single winner, so an
  envelope carrying its own empty `findings` beside the real nested one no longer ends the scan
  early.
- `recover_findings_json` stays byte-identical across both scripts, per the existing parity test in
  `tests/cross-model-recover-findings-parity.test.ts`.

`ce-pov`'s `recover_pov_json` already recurses into string values, so that path was never affected —
this brings the findings-shaped recovery to parity with it.

## Validation

Full suite: **2881 pass, 0 fail.**

Six new regression tests, five of them for failure modes an adversarial peer found in an earlier
draft of this patch (trailing non-findings object short-circuiting doc-review recovery; empty stub
trailing the real object; an envelope's own `findings: []` masking a nested review):

- `structuredOutput` (camelCase) is read, not just `structured_output`
- a populated review outranks an empty stub in either order
- recovery does not stop at an envelope's own empty `findings` beside `.text`
- recovery unwraps the string when jq cannot read the log at all
- a wrapped review survives grok's stub-then-real double emission
- (doc-review) a trailing non-findings object in `.text` does not short-circuit recovery

Replayed against the captured live envelope, the payload now resolves at the `.structuredOutput`
branch, and `recover_findings_json` alone still returns all 3 findings if every jq branch is
stripped.

## Not included

Two things I noticed and deliberately left alone, both behavior changes beyond this bug:

- When `recover_findings_json` finds a `findings` substring in the raw log but extracts nothing, that
  combination is strong evidence of an unwrapping gap rather than an empty peer, and arguably
  deserves a distinct log line so the orchestrator can report "parse failure, content present"
  instead of "no usable output".
- The pre-existing `.structured_output` and `.result` branches have the same unslurped-stream
  exposure as the `.text` one did, if a provider ever emits two envelopes.

## Reproducing

The full raw envelope is deleted by the skill's own post-fold-in cleanup, which is why the original
report of this had only a truncated log line to work from. Anyone reproducing should capture
`<run-dir>/jobs/<job-id>/out.log`, or tee the provider CLI's stdout, before the EXIT trap fires.

## Security Disclosure

No security-relevant changes. This touches JSON parsing of peer output only — no change to
egress, route selection, sandbox flags, prompt construction, path handling, or credentials. The
recovery scan's cost guard is widened from `"findings"` to `findings`, so a log merely containing
that word now enters the `raw_decode` scan; the scan remains bounded by the same substring test on
nested strings and a recursion depth cap of 3.

## Agent Disclosure

- **Model:** `Claude Code · claude-opus-5`

### CODE REVIEW

**Comment by tmchow** [comment]
## Review follow-up (empty `structuredOutput` vs `.text`)

I reviewed this for legitimacy and fix quality. Everything looks good, but i pushed 1 hardening commit onto this branch.

### Independent validation of the root cause

Reproduced against live `grok --json-schema --output-format json` on this machine. The envelope keys match the report:

- `structuredOutput` present (camelCase object) — **not** `structured_output`
- `text` holds the same payload as a **JSON string**
- Observed top-level keys: `text`, `stopReason`, `sessionId`, `requestId`, `usage`, `num_turns`, `total_cost_usd`, `total_cost_usd_ticks`, `modelUsage`, `structuredOutput`

So the original diagnosis was correct (good catch!): snake_case-only probing drops a complete grok peer review as “no usable output,” and recovery could not see `"findings"` nested inside a string.

### Residual Codex P2 — fixed in `c0e5e2ba`

Codex noted that:

```bash
jq -e '.structuredOutput | select((.findings|type)=="array")'
```

matches **empty** `findings: []` and returns before `.text` is checked. Confirmed with jq: empty camelCase still exits 0.

**Change:** both workers now probe in this order:

1. **Populated** `structuredOutput` (`findings` array length > 0)
2. **`.text`** (slurped, shape-filtered, prefer populated over empty — unchanged)
3. **Empty-but-shaped** `structuredOutput` — still a legitimate “peer found nothing,” only after `.text` had nothing better

Regression tests added:

- `empty structuredOutput does not preempt a populated .text review`
- `empty structuredOutput alone is still a zero-finding review`

### Not changed (still deliberate)

- Distinct log line for “`findings` substring present but extract failed” — still out of scope
- Pre-existing unslurped `.structured_output` / `.result` multi-object exposure — unchanged
- `ce-pov` fast path still snake_case-only; its string-recursive recovery already covers the nested case differently

Thanks for the careful diagnosis and the test suite around the envelope edge cases — this was a real silent fold-in miss on the grok-cli route!

File: skills/ce-code-review/scripts/cross-model-adversarial-review.sh [hunk 1]
```diff
@@ -732,6 +753,14 @@ parse_structured() {   # <logfile> <outfile>
   # Buffered single-object envelopes (grok-cli json, test stubs).
   jq -e '.structured_output | select((.findings|type)=="array")' "$1" > "$2" 2>/dev/null && return 0
   jq -r '.result // empty' "$1" 2>/dev/null | jq -e 'select((.findings|type)=="array")' > "$2" 2>/dev/null && return 0
+  # grok-cli names its parsed structured output in camelCase, so the snake_case
+  # probe above never matches it and a complete review looks like no output.
+  jq -e '.structuredOutput | select((.findings|type)=="array")' "$1" > "$2" 2>/dev/null && return 0
```
[the remark this answers is not in the document]
Comment (tmchow) [reply skills/ce-code-review/scripts/cross-model-adversarial-review.sh]: > When Grok's envelope contains an empty schema stub in `structuredOutput` while the verbatim `.text` stream contains the populated review, this new branch returns the empty array immediately

Fixed in c0e5e2ba — both workers now prefer populated `structuredOutput` first, then `.text` (populated-over-empty), then empty-but-shaped `structuredOutput` only as a legitimate zero-finding outcome. Regression tests cover empty-stub-does-not-preempt-text and empty-alone-still-publishes.

File: skills/ce-code-review/scripts/cross-model-adversarial-review.sh [hunk 2]
```diff
@@ -699,28 +701,47 @@ recover_findings_json() {   # <logfile> <outfile>
...
+            if depth < 3:
+                for v in obj.values():
+                    if isinstance(v, str) and 'findings' in v:
+                        scan(v, depth + 1)
+        i = end
+
+scan(txt, 0)
+# Last shaped object wins, but a populated review beats an empty one: grok emits
+# an empty stub beside the real object and the order is not guaranteed.
+best = next((o for o in reversed(found) if o["findings"]), found[-1] if found else None)
```
[the remark this answers is not in the document]
Comment (tmchow) [reply skills/ce-code-review/scripts/cross-model-adversarial-review.sh]: > if the peer emits an earlier draft object with findings and then its final schema object has `findings: []`, this line now walks backward to the first non-empty array and publishes the draft instead of the last terminal object

Fixed in 2c18f073 — recovery now tracks scan depth. Populated-over-empty applies only to string-unwrapped (nested `.text`) candidates where stub order is not guaranteed. Top-level sequential objects keep last-shaped-wins so a terminal `findings: []` after a draft is not revived. Regression covers the draft-then-empty case.
